### ISSUE TITLE
Please consider exporting DeepPartial

### ISSUE THREAD

**Issue by hwallis93** [stephenh/ts-proto#168]
We make heavy use of the `fromPartial` method, which takes a `DeepPartial` as its parameter. `DeepPartial` is defined in the autogenerated code, but not exported.

We have various functions that interact with `fromPartial`, meaning that we have various objects we would like to assign types to using `DeepPartial`. They might look something like this:
```typescript
function sendUpdate<ComplexType>(partialData: DeepPartial<ComplexType>) {
  const protobufForm = tsProtoBindings.ComplexType.fromPartial(partialData)
  sendToServer(protobufForm)
}
```

Currently we have to copy the definition of `DeepPartial` into our code, meaning that we have to make sure our definition and `ts-proto`'s defintion stay in sync. This isn't too onerous since it's a compile time error, but since it might be quite easy to simply export it I thought I'd ask

**Comment by ngrok-sarid** [comment]
Yea, we have to copy it ourselves as well. It looks like the following is copied to each output file:
```typescript
type Builtin = Date | Function | Uint8Array | string | number | undefined;
type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;
```

So, it brings up another question: Should all these common types go into a single file that gets exported/imported or do you simply export the type in each file (which would have it's own problems, maybe?).

**Comment by ngrok-sarid** [comment]
@hwallis93 It's not pretty, but you can also do this:
```typescript
import { MyService } from "...wherever";
type ArgType = Parameters<MyService["MyMethod"]>[0];
```

**Comment by stephenh** [comment]
I'm good with `export`ing `DeepPartial`. There is a small caveat that, as written, it may not work against every TS type ever, i.e. I wouldn't drop it into a utils library as the most-perfect `DeepPartial` impl ever. But it works against the subset of TS types that ts-proto generates based on proto files.

So, that disclaimer aside.

Also the reason that it's generated in every file was to avoid the `ts-proto` code having any runtime dependencies. I.e. we'd need to start having some sort of `ts-proto-utils` or `ts-proto-runtime` node module that would define that and then all projects would have to pull in as a dependency. Which, dunno, maybe we'll end up with that at some point, but it hasn't seemed worth the trade-off yet.

**Comment by stephenh** [comment]
Fixed in 1.39.0.

**Comment by hwallis93** [comment]
Thanks very much @stephenh . We currently `import *` from a collection of files so we'll have to find a way around the name clash but I'm sure that's solvable.
